### PR TITLE
WIP: Authorization for release aggregates

### PR DESCRIPTION
**Pull request by fbertsch**

### CODE REVIEW

File: mozaggregator/service.py [hunk 1]
```diff
@@ -212,18 +327,33 @@ def status():
...
 @add_cache_header()
 @cache_request
 def get_channels(prefix):
     channels = execute_query("select * from list_channels(%s)", (prefix, ))
-    return Response(json.dumps([channel[0] for channel in channels]), mimetype="application/json")
+    channels = [channel[0] for channel in channels]
+
+    try:
+        check_auth()
+    except AuthError:
```
Comment (chutten) [inline mozaggregator/service.py]: Not a big fan of using exceptions for non-exceptional cases

File: mozaggregator/service.py [hunk 2]
```diff
@@ -2,15 +2,15 @@
...
 import boto3
 import ujson as json
 from botocore.exceptions import ClientError
 from dockerflow.flask import Dockerflow
-from flask import Flask, Response, abort, request
-from flask.ext.cache import Cache
-from flask.ext.cors import CORS
+from flask import Flask, Response, abort, jsonify, request, _request_ctx_stack
+from flask_cache import Cache
+from flask_cors import CORS
```
Comment (robhudson) [inline mozaggregator/service.py]: I think that means the errors locally will go away!! Yay!!

File: mozaggregator/service.py [hunk 3]
```diff
@@ -68,6 +70,119 @@
 ALLOW_ALL_RELEASE_METRICS = os.environ.get("ALLOW_ALL_RELEASE_METRICS", "False") == "True"
 PUBLIC_RELEASE_METRICS = {"SCALARS_TELEMETRY.TEST.KEYED_UNSIGNED_INT"}
 
+# Auth0 Integration
+AUTH0_DOMAIN = "chutten.auth0.com"
+API_AUDIENCE = "aggregates.telemetry.mozilla.org"
+ALGORITHMS = ["RS256"]
+REQUIRED_SCOPE = "read:aggregates"
```
Comment (robhudson) [inline mozaggregator/service.py]: I wonder if we can prefix these with `AUTH0` to help signify their purpose later in the code?

File: mozaggregator/service.py (hunk 1 above)
Comment (fbertsch) [reply mozaggregator/service.py]: I know it feels odd, but it's a real thing: https://blogs.msdn.microsoft.com/pythonengineering/2016/06/29/idiomatic-python-eafp-versus-lbyl/

File: mozaggregator/service.py (hunk 2 above)
Comment (fbertsch) [reply mozaggregator/service.py]: Right!

File: mozaggregator/service.py (hunk 3 above)
Comment (fbertsch) [reply mozaggregator/service.py]: That seems like a good idea.

**Comment by fbertsch** [comment]
@chutten @robhudson want to take another look before I merge?

File: mozaggregator/service.py [hunk 4]
```diff
@@ -59,22 +61,156 @@
 # Disallowed metrics for serving
 METRICS_BLACKLIST = [
     "SEARCH_COUNTS",
```
Comment (chutten) [inline mozaggregator/service.py]: Of interest: The first thing mreid looked for was SEARCH_COUNTS... we may be asked to add that on for auth'd clients.

File: mozaggregator/service.py [hunk 5]
```diff
@@ -59,22 +61,156 @@
...
+    """
+    auth = request.headers.get("Authorization", None)
+    if not auth:
+        raise AuthError({"code": "authorization_header_missing",
+                        "description":
+                            "Authorization header is expected"}, 403)
+
+    parts = auth.split()
+
+    if parts[0].lower() != "bearer":
```
Comment (chutten) [inline mozaggregator/service.py]: Do we want to strictly match "Bearer" with specific caps?

File: mozaggregator/service.py [hunk 6]
```diff
@@ -59,22 +61,156 @@
...
+    jwks = json.loads(jsonurl.read())
+
+    try:
+        unverified_header = jwt.get_unverified_header(token)
+    except JWTError:
+        raise AuthError({"code": "improper_token",
+                        "description": "Token cannot be validated"}, 403)
+
+    rsa_key = {}
+    for key in jwks["keys"]:
```
Comment (chutten) [inline mozaggregator/service.py]: So whichever key's id matches the header's key id _last_ wins out? Or do we only expect one matching id? (And if so, should we break the loop at that point?)

File: mozaggregator/service.py [hunk 7]
```diff
@@ -59,22 +61,156 @@
...
+        if unverified_claims.get("scope"):
+            token_scopes = unverified_claims["scope"].split()
+            for token_scope in token_scopes:
+                if token_scope == AUTH0_REQUIRED_SCOPE:
+                    _request_ctx_stack.top.current_user = payload
+                    auth0_cache[token] = True
+                    return True
+            raise AuthError({"code": "access_denied",
+                            "description": "Access not allowed"}, 403)
+
```
Comment (chutten) [inline mozaggregator/service.py]: We have one last error case here for unverified claims without a scope which would fall through and claim to be "Unable to find appropriate key"

File: mozaggregator/service.py [hunk 8]
```diff
@@ -59,22 +61,156 @@
...
+                audience=AUTH0_API_AUDIENCE,
+                issuer=domain_base
+            )
+        except jwt.ExpiredSignatureError:
+            raise AuthError({"code": "token_expired",
+                            "description": "token is expired"}, 403)
+        except jwt.JWTClaimsError:
+            raise AuthError({"code": "invalid_claims",
+                            "description":
+                                "incorrect claims,"
```
Comment (chutten) [inline mozaggregator/service.py]: Inconsistent capitalization in error messages. "incorrect claims..." and "token is expired" seem to be the only two without an initial capital.

File: mozaggregator/service.py (hunk 5 above)
Comment (fbertsch) [reply mozaggregator/service.py]: Seems reasonable.

File: mozaggregator/service.py (hunk 6 above)
Comment (fbertsch) [reply mozaggregator/service.py]: Right, just one match on `kid`. I went ahead and simplified this with a `for/else` instead, so if we exit that block without raising an error we are guaranteed to have an `rsa_key`.

File: mozaggregator/service.py (hunk 7 above)
Comment (fbertsch) [reply mozaggregator/service.py]: Okay, now if scope fails at all (but we found the `rsa_key`), it will return with `Access not allowed`. Seem reasonable?

**Pull request merged**